course_contents/2_intro_to_python/notes/02. Loops.md

```diff
@@ -0,0 +1,159 @@
+```
+
+### The `for` loop
+- for loop runs a given number of times
+- general syntax for `for` loop is:
+
+```py
+for variable in iterable:
+    block_of_code
+```
+- an iterable is an object that can be iterated over e.g. a list
+- an object is iterable if it can return the iterator object (more on that in the OOP section)
+
+An example of the `for` loop:
+
+```py
+numbers = [1,2,3,4,5]
+for i in numbers:
+    print(i)
+```
+
+**Extra resources**
+[How does the for loop work in Python?](https://python.tecladocode.com/3_continuous_countries/1_loops.html#how-does-the-for-loop-work-in-python)
+
+#### The `range()` method
+- the `range()` method returns a sequence of integers in a given range
+- arguments that are passed to the method are:
+    1. `start` - an integer marking the start of a sequence
+    2. `stop` - an integer marking the end of a sequence. Important - `range()` returns integers up to `stop-1`
+    3. `step` - an integer representing the size of the step i.e. the difference between consecutive integers in a sequence
+- the numbers are generated by starting from `start` and incrementing it by `step` until `stop-1` is reached
+- the only argument that has to be stated explicitly is the `stop`
+- if not specified, Python will implicitly set `start` to 0 and `step` to 1
+
+An example of the `range()` method used in a `for` loop:
+
+```py
+for i in range(5): #the range() method can be used to specify number of iterations of a loop
+  print(i) #the numbers printed will be: 0,1,2,3,4
+```
+
+### The `break` statement
+- the `break` statement is used to stop the loop from running when an external condition is reached
+- after the loop has been exited - the next line of code after it will be executed
+
+An example of using the `break` statement in the `while` loop:
+
+```py
+counter = 0
+while counter < 10:
+  print(counter)
+  if counter == 5:
+    break #the loop stops running when the counter is equal to 5
+  counter += 1
+print(f"That's all the numbers from 0 to {counter}!") #after the break, this line is executed
+```
+
+### The `continue` statement
+- the `continue` statement is used to skip a step in a loop and run the next iteration
+- no code after the `continue` statement will be run
+
+An example of the `continue` statement being used in a `for` loop:
+
+```py
+for i in range(5):
+    if i % 2 == 0: 
+        continue #if i is even, it won't be printed out
+    print(i)
+```
+
+### The `pass` statement
+- the `pass` statement is used to ignore a place in a code and continue running as if it doesn't exist
+- it is usually used as a placeholder at a place where some code is required from the syntax side, but the code doesn't exist yet
+
+An example of the `pass` statement being used in a `for` loop:
+
+```py
+for i in range(5):
+    if i % 2 == 0: 
+        pass #even if the block is entered nothing will happen because of the pass statement
+    print(i)
+```
+
+### `else` keyword and `for` loop
```

Comment:
I would reword this to `else` in `for` loops, or something similar, but not `and`